### ISSUE TITLE
Can't get exceptions to work

### ISSUE THREAD

**Issue by mre** [PyO3/pyo3#171]
### 🌍 Environment

 - Your operating system and version: MacOS X 10.12
 - Your python version: 3.6.5
 - How did you install python (e.g. apt or pyenv)? Did you use a virtualenv?: brew, no virtualenv
 - Your rust version: rustc 1.28.0-nightly (5bf68db6e 2018-05-28)

### 💥 Reproducing

I somehow can't exceptions to work in my project.

To reproduce the issue, I simple hacked an exception to the word-count example code.
Here's the diff of my changes: https://github.com/PyO3/pyo3/compare/master...mre:master

Here is a simple test case to reproduce my issue.

```
cd /tmp
git clone git@github.com:mre/pyo3.git
cd pyo3/examples/word-count
python3 setup.py install
cd ~
python3 -c "from word_count import search"
```

I expected a `JSONDecodeError`, but got this:

```
python3 -c "from word_count import search"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.6/site-packages/word_count-0.1.0-py3.6-macosx-10.12-x86_64.egg/word_count/__init__.py", line 4, in <module>
    from ._word_count import search, search_sequential, JSONDecodeError
ImportError: cannot import name 'JSONDecodeError'
```

Any help would be highly appreciated.

**Comment by konstin** [comment]
What you did is defining your own `JSONDecodeError` in rust, which, when returned to python code, raises this custom exception. You can't import that as an exception class in python since you didn't add it to module, which is the reason for the import error.

What I believe you actually want is to import the `JSONDecodeError` from the python's std lib both in rust and in python.

**Comment by mre** [comment]
Oh, I thought I add it to the module like [this](https://github.com/PyO3/pyo3/compare/master...mre:master#diff-40d3d6865379930bee4b1af0e7197d75)? Apparently that's not enough.

> What I believe you actually want is to import the JSONDecodeError from the python's std lib both in rust and in python.

Good idea. My only concern is, that `JSONDecodeError` is part of the json module and [what I'm trying to build](https://github.com/mre/hyperjson) aims to be a drop-in replacement for that. I'm afraid of this direct dependency, that's why I wanted to declare my own.
That said, it might not even matter as `JSONDecodeError` will always be available for import since it is part of the stdlib. So it should be fine, no?

**Comment by konstin** [comment]
> Oh, I thought I add it to the module like this? Apparently that's not enough.

py_exception is only a shortcut to create a unit struct and adding a bunch of impls. The module name is actually just stringified, so could pass any conceivable name there, whether that name exists or not ([source](https://docs.rs/pyo3/0.2.6/src/pyo3/err.rs.html#47-107) for reference). `py.get_type::<JSONDecodeError>()` gives you an object that you can add to module like any other value.

> Good idea. My only concern is, that JSONDecodeError is part of the json module and what I'm trying to build aims to be a drop-in replacement for that. I'm afraid of this direct dependency, that's why I wanted to declare my own.

> That said, it might not even matter as JSONDecodeError will always be available for import since it is part of the stdlib. So it should be fine, no?

Well those things are in the standard library so you don't have extra dependencies; Whether you want to reuse stdlib or build you're own is a design choice you have to make for your project.

**Comment by mre** [comment]
Thanks! I tried to expose the exception now like this ([change here](https://github.com/mre/pyo3/commit/b4cb1775f213292f8390b82bfd26467624f80374#diff-866c0cb8fa85d853a049969541321fa1R58)):

```rust
m.add("JSONDecodeError", py.get_type::<JSONDecodeError>());
```

But now when I run it, I get a panic.

```
python3 -c "from word_count import search"
thread '<unnamed>' panicked at 'assertion failed: `(left != right)`
  left: `0`,
 right: `0`', /Users/mendler/.cargo/registry/src/github.com-1ecc6299db9ec823/pyo3-0.2.7/src/pythonrun.rs:42:13
note: Run with `RUST_BACKTRACE=1` for a backtrace.
fatal runtime error: failed to initiate panic, error 5
```

I'm pretty sure I'm doing it wrong, but I guess it should still not panic here, right? 😅 

Also, is there an example snippet for exposing the exception correctly? I was looking around but couldn't find anything yet. Thanks again for your help.

**Comment by mre** [comment]
Currently I use the mentioned solution to import `json.JSONDecodeError` with `import_exception!(json, JSONDecodeError);` into my Rust code and then using it like so:

```rust
Err(JSONDecodeError::new((
  format!("Value: {:?}, Error: {}", s, e),
  s.to_object(py),
  0,
)))
```

Not beautiful, but it works. Would still be curious to know how to fix the issue mentioned in my previous comment, but for now it's okay. For the sake of keeping the issue count low, I'll close this for now. Feel free to reopen in case you'd like to add something.
Thanks to @konstin for providing guidance. 👍

**Comment by konstin** [comment]
I've opened #177 because that panic definitely should not happen.

Regarding an example for exceptions: There is currently only https://pyo3.rs/master/exception.html and lack of unit tests for exceptions, which leads to problems such as the bug you ran into. You'll probably have most success reading the source and sometimes even the python guide for writing C/C++ extensions, fixing things as you're building upon them (which is how I got into this project).
